src/libutil/thread.cpp

```diff
@@ -49,15 +49,72 @@
 #include <future>
 #include <memory>
 
+#include <OpenImageIO/sysutil.h>
+#include <OpenImageIO/thread.h>
+#include <OpenImageIO/sysutil.h>
+#include <OpenImageIO/strutil.h>
+
+#include <boost/thread/tss.hpp>
+
+#if 0
+
 // Use boost::lockfree::queue for the task queue
 #include <boost/lockfree/queue.hpp>
+template<typename T> using Queue = boost::lockfree::queue<T>;
 
-#include <OpenImageIO/sysutil.h>
-#include <OpenImageIO/thread.h>
+#else
+
+#include <queue>
+
+namespace {
+
+template <typename T>
+class Queue {
+public:
+    Queue (int size) { }
+    bool push(T const & value) {
+        std::unique_lock<Mutex> lock(this->mutex);
+        this->q.push(value);
+        return true;
+    }
+    // deletes the retrieved element, do not use for non integral types
+    bool pop(T & v) {
+        std::unique_lock<Mutex> lock(this->mutex);
+        if (this->q.empty())
+            return false;
+        v = this->q.front();
+        this->q.pop();
+        return true;
+    }
+    bool empty() {
+        std::unique_lock<Mutex> lock(this->mutex);
+        return this->q.empty();
+    }
+private:
+    typedef OIIO::spin_mutex Mutex;
+    std::queue<T> q;
+    Mutex mutex;
+};
+
+}
+
+#endif
 
 
 OIIO_NAMESPACE_BEGIN
 
+bool debug_threads = false; // FIXME -- remove this after debugging
+
+static int
+threads_default ()
+{
+    int n = Strutil::from_string<int>(Sysutil::getenv("OPENIMAGEIO_THREADS"));
+    if (n < 1)
```

Comment:
There are two different concepts: How many threads we want in the app, and how many workers in the thread pool.

If you don't want to "overthread", you probably want the pool to be cores-1, because you already have one thread putting things into the work queue and performing some of the work itself.

Number of default threads is equal to number of cores. It doesn't make sense to have 0 threads, because something needs to run. So requesting anything < 1 means "use the default", which is number of cores (unless overridden by OPENIMAGEIO_THREADS, which I mainly use for debugging).

The *pool* size, however, is a different matter. There, it's reasonable to request 0, which means that the calling thread should do all the work, and there are no extra threads just watching the pool's work queue. The pool will default to threads_default-1.

Make sense?